### PR TITLE
(feat) O3-3423: Allow user to change password

### PR DESCRIPTION
**Pull request by ibacher**
# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [ ] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

### CODE REVIEW

File: packages/apps/esm-login-app/src/change-password/change-password.extension.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import { Button, SwitcherItem } from '@carbon/react';
+import styles from './change-password.scss';
+
+const ChangePasswordLink: React.FC = () => {
+  const { t } = useTranslation();
+
+  const launchChangePasswordModal = useCallback(() => showModal('change-password-modal'), []);
+
+  return (
+    <SwitcherItem aria-label="Change Password" className={styles.panelItemContainer}>
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.extension.tsx:13]: This aria-label should be translated.

File: packages/apps/esm-login-app/src/change-password/change-password.extension.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      <div>
+        <UserAvatarIcon size={20} />
+        <p>Password</p>
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.extension.tsx:16]: Ditto here.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import React, { type ChangeEvent, useCallback, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { Button, InlineLoading, ModalBody, ModalFooter, ModalHeader } from '@carbon/react';
+import styles from './change-password-modal.scss';
+import { Form, PasswordInput } from '@carbon/react';
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:5]: Consolidate the `@carbon/react` imports into one line.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  return (
+    <>
+      <ModalHeader closeModal={close} title={t('changePassword', 'Change password')} />
+      <ModalBody>
+        <div className={styles.languageOptionsContainer}>
+          <Form>
+            <PasswordInput
+              onChange={(event: ChangeEvent<HTMLInputElement>) => setOldPassword(event.target.value)}
+              labelText={t('oldPassword', 'Old Password')}
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:30]:
```suggestion
              labelText={t('oldPassword', 'Old password')}
```

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            />
+            <PasswordInput
+              onChange={(event: ChangeEvent<HTMLInputElement>) => setNewPassword(event.target.value)}
+              labelText={t('newPassword', 'New Password')}
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:34]:
```suggestion
              labelText={t('newPassword', 'New password')}
```

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx (hunk 4 above)
Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:30]: Prefer [sentence-case](https://carbondesignsystem.com/guidelines/content/writing-style/#use-sentence-case-capitalization).

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            />
+            <PasswordInput
+              onChange={(event: ChangeEvent<HTMLInputElement>) => setConfirmPassword(event.target.value)}
+              labelText={t('confirmPassword', 'Confirm New Password')}
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:38]:
```suggestion
              labelText={t('confirmPassword', 'Confirm new password')}
```

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import React, { type ChangeEvent, useCallback, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { Button, InlineLoading, ModalBody, ModalFooter, ModalHeader } from '@carbon/react';
+import styles from './change-password-modal.scss';
+import { Form, PasswordInput } from '@carbon/react';
+
+interface ChangePasswordModalProps {
+  close(): void;
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:8]:
```suggestion
  close: () => void;
```

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+              labelText={t('confirmPassword', 'Confirm New Password')}
+            />
+          </Form>
+        </div>
+      </ModalBody>
+      <ModalFooter>
+        <Button kind="secondary" onClick={close}>
+          {t('cancel', 'Cancel')}
+        </Button>
+        <Button className={styles.submitButton} disabled={isChangingPassword} type="submit" onClick={handleSubmit}>
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:47]: You probably want to have the following CSS for the loading state in `change-password-modal.scss`:

```scss
.submitButton {
  :global(.cds--inline-loading) {
    min-height: 1rem;
  }

  :global(.cds--inline-loading__text) {
    font-size: unset;
  }
}
```

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  const handleSubmit = useCallback(() => {
+    setIsChangingPassword(true);
+  }, []);
+
+  return (
+    <>
+      <ModalHeader closeModal={close} title={t('changePassword', 'Change password')} />
+      <ModalBody>
+        <div className={styles.languageOptionsContainer}>
+          <Form>
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:27]: You probably want to wrap these password inputs in a [Stack](https://react.carbondesignsystem.com/?path=/docs/layout-stack--overview) with a gap prop set to something like `5` for consistent spacing.

**Review by denniskigen — COMMENTED** [review]
Good start. I've left some feedback.

File: packages/apps/esm-login-app/src/change-password/change-password.extension.tsx (hunk 2 above)
Comment (gitcliff) [reply packages/apps/esm-login-app/src/change-password/change-password.extension.tsx]: thanks @denniskigen ,,,
have addressed the comments

**Review by ibacher — COMMENTED** [review]
Thanks @gitcliff! I think some more work is needed here on validations. Before submitting to the backend, we should ensure that the `oldPassword` has a non-empty value. Also if the `newPassword` is empty, the error message should be something like "New password is required". We might also try setting the `required` prop on all three fields.

File: packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React, { useCallback } from 'react';
+import { useTranslation } from 'react-i18next';
+import { Password } from '@carbon/react/icons';
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx:3]: We can revert to using `PasswordIcon` from `@openmrs/esm-framework` in a future commit. Making the change here causes some errors.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    required_error: t('oldPasswordRequired', 'Old password is required'),
+  });
+
+  const newPasswordValidation = z.string({
+    required_error: t('newPasswordRequired', 'New password is required'),
+  });
+
+  const passwordConfirmationValidation = z.string({
+    required_error: t('passwordConfirmationRequired', 'Password confirmation is required'),
+  });
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:29]: This Zod schema validates all three inputs and automatically displays custom error messages if any inputs are missing.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  const changePasswordFormSchema = z
+    .object({
+      oldPassword: oldPasswordValidation,
+      newPassword: newPasswordValidation,
+      passwordConfirmation: passwordConfirmationValidation,
+    })
+    .refine((data) => data.newPassword === data.passwordConfirmation, {
+      message: t('passwordsDoNotMatch', 'Passwords do not match'),
+      path: ['passwordConfirmation'],
+    });
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:40]: This `refine` block ensures the new password and password confirmation values match, displaying an error if they do not.

File: packages/apps/esm-login-app/src/routes.json [hunk 13]
```diff
@@ -32,6 +32,13 @@
       "online": true,
       "offline": true
     },
+    {
+      "name": "password-changer",
+      "slot": "user-panel-slot",
+      "component": "changePasswordLink",
+      "online": true,
+      "offline": true
```
Comment (denniskigen) [inline packages/apps/esm-login-app/src/routes.json:40]: Renamed to align with the naming convention for similar components.

**Comment by denniskigen** [comment]
@gitcliff I've pushed a commit that adds validation using a Zod schema. Please test it and confirm that it works as expected.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx (hunk 12 above)
Comment (ibacher) [reply packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:40]: I think the `path` here should be `['passwordNew', 'passwordConfirmation']`, i.e., it should mark both fields as errors, no?

**Comment by denniskigen** [comment]
@ibacher, should the password change link be controlled by user privilege?

**Comment by ibacher** [comment]
@denniskigen No. This just allows user's to change their own passwords, which, IMO, everyone should be able to do.

File: packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx (hunk 10 above)
Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx]: Specifically this, @ibacher:

CleanShot 2024-07-02 at 3  28 48@2x

CleanShot 2024-07-02 at 3  29 56@2x

Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx]: Or maybe I should just rerun the styleguide build to fix it?

Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx]: That did the trick.

Comment (ibacher) [reply packages/apps/esm-login-app/src/change-password/change-password-link.extension.tsx]: Ah... sorry, yeah, I've been using `run:shell` and import map overrides to test this out. It was causing an issue with the tests though. I wish I had a better system for the icons, honestly.

File: packages/apps/esm-login-app/src/change-password/change-password.modal.tsx (hunk 12 above)
Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:40]: AFAICT, Zod doesn't support assigning an error message to multiple fields directly. To show error messages on both fields, we need to set them manually. Typically, the error message appears only on the confirmation field.

Comment (denniskigen) [reply packages/apps/esm-login-app/src/change-password/change-password.modal.tsx:40]: https://github.com/openmrs/openmrs-esm-core/assets/8509731/88e689af-3505-4ba4-8235-e338be858028

**Comment by ibacher** [comment]
@denniskigen Could you approve this? Since I opened the PR I can't, but this LGTM.

**Review by denniskigen — APPROVED** [review]
LGTM

**Pull request merged**